### ISSUE TITLE
Consecutive calls to promise.future() return different instances when debugging

### ISSUE THREAD

**Issue by ptomaszek** [vert-x3/vertx-lang-groovy#113]
### Version

3.9.3

### Context
```groovy
Promise p = Promise.promise()
println p.future() == p.future() // prints false when debugging
```

Instances are not equal when running code in the Debug mode under Intellij IDEA.

### Steps to reproduce

1. Code: https://github.com/ptomaszek/vertx-groovy
1. Run `MainGroovy` -> observe expected output `true`
1. Run in Debug mode `MainGroovy` -> observe unexpected output `false`

### Extra

 - IntelliJ IDEA 2020.2.3
 - Code works as expected with `4.0.0-milestone5`
 - There's `MainJava` for a reference showing that plain Java works as expected with both Run and Debug
 - JVM:
   ```
    java version "1.8.0_201"
    Java(TM) SE Runtime Environment (build 1.8.0_201-b09)
    Java HotSpot(TM) 64-Bit Server VM (build 25.201-b09, mixed mode)
   ```

**Comment by ptomaszek** [comment]
I also checked with Eclipse 2020-09:
 - 3.9.3 - `false` in both Run and Debug
 - 4.0.0-milestone5 - `true` in both Run and Debug

**Comment by tsegismont** [comment]
I'm not sure why you get a different behavior in run/debug mode with `3.9.3`, it should be the same.

However I can explain the difference between `3.9.3` and `4.0.0.milestone5`: in Vert.x 4, the `Promise` implementation has its own class, while in Vert.x 3 it's shared with `Future`. Consequently, when Groovy resolves the `future()` method call dynamically, it invokes the `Future.future()` static method instead of the instance `Promise.future()` method.

If you annotate the class with `@CompileStatic`it will work as expected.

I'm not very knowledgeable with Groovy, any idea @vietj about this one?

**Comment by aaloise** [comment]
The answer is probably [here](https://mrhaki.blogspot.com/2009/09/groovy-goodness-check-for-object.html?m=1#:~:text=In%20Groovy%20we%20use%20the,The%20!%3D) 

"Groovy overloads the == operator and maps it to the equals() method. This is very different from Java, so when developers are switching back and forth between Groovy and Java mistakes are bound to happen. In Java we use the == operator to see if variables are referring to the same object instance. In Groovy we use the == operator to see if two objects are the same, in Java we would use the equals() method for this. To test if two variables are referring to the same object instance in Groovy we use the is() method. The != operator is also overloaded and maps to the !equals() statement."

**Comment by tsegismont** [comment]
@aaloise it's not a problem with the operator, it's actually a bug.

The `Future` returned by calling `promise.future()` in Groovy with Vert.x 3.9 is actually not bound to the `Promise`

Consider this program:

```groovy
Promise p = Promise.promise()
p.future().onSuccess { println it }
p.complete("Done")
```

If you run it with Vert.x 3 you'll see nothing, but with Vert.x 4 you'll see `Done` in the console, as expected.

**Comment by ptomaszek** [comment]
Thank you guys for quick responses.
And especially @tsegismont for nicely describing the actual issue.

So I guess it's not safe to use `Promise`s and best to stick to deprecated but working `def f = Future.future()` for time being.

Fingers crossed for a bugfix :crossed_fingers:

**Comment by tsegismont** [comment]
@ptomaszek Vert.x internals work with promises in Java, and only in Vert.x 4

**Comment by ptomaszek** [comment]
@aaloise, @tsegismont 
Did you get a chance to look into what's causing the issue?

**Comment by aaloise** [comment]
Well, I noticed that ScriptVerticle in 3.9 is still using Future. In 4.0 it uses Promise

https://github.com/vert-x3/vertx-lang-groovy/blob/3.9/vertx-lang-groovy-gen/src/main/java/io/vertx/lang/groovy/ScriptVerticle.java (3.9)

https://github.com/vert-x3/vertx-lang-groovy/blob/master/src/main/java/io/vertx/lang/groovy/ScriptVerticle.java (4.0)

Maybe it's related? WDYT @tsegismont @vietj

**Comment by aaloise** [comment]
I submitted a PR to fix ScriptVerticle (#114). Still don't know if it is related, but either way I believe it is necessary.

In addition, I did new tests. In a main method I have the following line codes

```
Promise p = Promise.promise()
println p.future() == p.future()
		
Promise p2 = Promise.promise()
p2.future().onSuccess { println it }
p2.complete("Done!!!")
````

So it should always print:
```
true
Done!!!
```

But it show only false. The I decide to run it 10 times, and to my surprise it printed false 5 times and true with Done the other 5 times.

I noticed the same behaviour using a verticle. But it only happens with Groovy

**Comment by aaloise** [comment]
It seems that the bug appeared in version 3.8.0 with the arise of Promise. When it was just Future, for obvious reasons, the problem did not occur.

I noticed that Promise's Future has the @CacheReturn annotation, which states: "Declares that the value returned by a method will always be the same. Code generators can use this to cache the proxy of the returned object and avoid to recreate a different proxy for the same object."

So I was wondering, can this be related to codegen.....?

**Comment by vietj** [comment]
@aaloise generators can use it, however we don't use proxy anymore for groovy.

do you have a reproducer ?

**Comment by aaloise** [comment]
@vietj I've been using an adapted @ptomaszek reproducer to do some tests, with suggestions made by @tsegismont. Basically:

```
package org.example


import io.vertx.core.Promise

class MainGroovy {

    static void main(String[] args) {
        Promise p = Promise.promise()
        println p.future() == p.future()

        Promise p2 = Promise.promise()
        p2.future().onSuccess { println it }
        p2.complete("Done!!!")
    }
}
```

This should print ```true``` and ```Done!!!```, but is always printing only ```false```. I think I've found an workaround, which is annotating the class MainGroovy with ```@TypeChecked``` or ```@CompileStatic```

@ptomaszek could you give a try with one of this annotations? Like this:

```
package org.example


import io.vertx.core.*
import groovy.transform.*

@TypeChecked
class MainGroovy {

    static void main(String[] args) {
        Promise p = Promise.promise()
        println p.future() == p.future()

        Promise p2 = Promise.promise()
        p2.future().onSuccess { println it }
        p2.complete("Done!!!")
    }
}
```

**Comment by ptomaszek** [comment]
@aaloise, Thanks for diving into the issue.

Annotations you've mentioned work for simple classes like `MainGroovy`. However the project I'm working on has dozens of complex files that leverage Groovy's magic. Annotating these breaks compilation and eventual refactoring would be tremendously costly in time.

To sum up - unfortunately the workaround is not for me, and I think it's not a good suite for any ongoing product.

I really hope you could figure this out. I tried myself but with no effect. I thought the issue is caused by not handling `@CacheReturn` but I did a simple check that disapproved it, i.e. this always passes:

```
    def instance = Vertx.vertx()
    Assert.assertEquals(instance.fileSystem(), instance.fileSystem())
```

So the issue seems to be affecting `Promise` only.

**Comment by aaloise** [comment]
@ptomaszek I see. I too noticed that must be something with Promise.

One curious thing is that your reproducer always prints false with OpenJDK Hotspot. I tried sometimes with OpenJ9 and ocasionallly it printed true. Let's keep investigating.

**Comment by aaloise** [comment]
@ptomaszek If you try to run your reproducer with 3.9.5-SNAPSHOT, with embedded vertx in the code, seems to work as expected.

```
package org.example

import io.vertx.core.*

class MainGroovy {

    static void main(String[] args) {
        Vertx vertx = Vertx.vertx()
	Promise p = Promise.promise()
        println p.future() == p.future()

        Promise p2 = Promise.promise()
        p2.future().onSuccess { println it }
        p2.complete("Done!!!")
    }
}
```

When I run this code using 3.9.5-SNAPSHOT, it prints ```true``` and ```Done!!!```. This seems to be related with your last post, but as I mentioned, this peace of code only prints the expected wiht 3.9.5-SNAPSHOT.

The "digging" continues...

**Comment by ptomaszek** [comment]
I wasn't sure whether you're referring to  3.9.5-SNAPSHOT in `vertx-lang-groovy` or `vertx-core` so I downloaded sources for both, compiled and installed locally, and then configured gradle's `dependencies`:
```
    implementation 'io.vertx:vertx-core:3.9.5-SNAPSHOT'
    implementation 'io.vertx:vertx-lang-groovy:3.9.5-SNAPSHOT'
```

Still no luck, i.e. random results.

---

I just noticed that with different JVMs issue repeatability is different, e.g. for one JVM consecutive Run/Debug operations produce random results, and in another each Run produces same result, but Debug opposite.

I know it's not really about Run/Debug but issue repeatability is really somehow connected with how JVM operates.

Also I noticed that if you build a fat jar and run the application with this code:

```groovy
class MainGroovy {

    static void main(String[] args) {

        (0..10).each {
            Promise promise = Promise.promise()
            println promise == promise.future()
        }
    }
}
```

then you'll either always get all the results `true` or `false`. Never mixed. So the problem is probably not related to Runtime, but some earlier JVM's step, like classloading, linking or initialization. But it still may be Runtime, I'm not an expert here.

Fat jar/shadowjar above started with `AdoptOpenJDK 11.0.7.hs` will randomly produce 10 `true` or 10 `false` values;
but started with `AdoptOpenJDK 8.0.265.hs` will always give `true`.

It's crazy.

**Comment by aaloise** [comment]
Did some tests also with that pice of code. Always with 3.9.5-SNAPSHOT. I'm getting all kind of results, either running or building a jar. Tried with four JVM's: 1.8.0.272 Hotspot; 1.8.0.272 OpenJ9; 11.0.9 Hotspot; 11.0.9 OpenJ9. All from AdoptJDK.

With JDK 8 Hotspot I only get a true as result using @CompileStatic. With JDK 11 Hostspot I see true with embedded Vertx or with @CompileStatic. With JDK 8 or 11 OpenJ9 I see alternate results: sometimes true, sometimes false, with or without embedding Vertx, but wiht @CompileStatic is always true.

To sum up. Comparing all four JVM's, apparently only when @CompileStatic is used, the results shows true.

I agree. Seems crazy. But has to be some logic behind this.

**Comment by aaloise** [comment]
@ptomaszek I tried to run the code with Groovy Console and works as expected:

![image](https://user-images.githubusercontent.com/1064972/98560097-c8cbfc00-227d-11eb-84cd-746ceb352f5f.png)

There must be somethig else messing around with the results

**Comment by aaloise** [comment]
I tried the following and seems to work. Created a file named groovyOptionsConfig.groovy with the following content.

```
import org.codehaus.groovy.ast.ClassNode


withConfig(configuration) {

    source(classValidator: { ClassNode cn -> cn.name == "io.vertx.core.Promise") }) {
        ast(groovy.transform.CompileStatic)
    }

}
```

And call it in build.gradle:

```
compileGroovy {
    groovyOptions.configurationScript = file("groovyOptionsConfig.groovy")
}
```

That way only Promise compiles with @CompileStatic

**Comment by ptomaszek** [comment]
@aaloise, Can you please share a complete example? AST to `Promise` doesn't work for me.

PS Closing bracket at `Promise")` fails compilation

**Comment by aaloise** [comment]
@ptomaszek my last post turned out not to be a good solution. I faced odd behaviours using this approach in different machines, even with the same JVM.

So when @tsegismont posted earlier, stating that the problem is when promise.future() gets called, I thought of creating a Helper class that compiles statically this method. Something like:

```
package org.example

import groovy.transform.*
import io.vertx.core.*

class FutureHelper {

    @CompileStatic
    static def future(Promise promise){
        promise.future()
    }
}
```

And then, your reproducer could be adapted to:

```
package org.example

import io.vertx.core.*
import org.example.FutureHelper as FH

class MainGroovy {

    static void main(String[] args) {
        Promise p = Promise.promise()
        println FH.future(p) == FH.future(p)

    }
}
```

I'm getting true everytime. I know that this is far from the ideal, but for now is the best I could achieve.

**Comment by ptomaszek** [comment]
@aaloise, thanks for the clever workaround but it requires a lot of refactoring and I'm not sure it's bullet-proof. When non-custom code - e.g. `Veticle.start(Promise<Void> startPromise)` - accepts `Promise`, then it'll still invoke `promise.future()` directly rather than with a helper.

**Comment by ptomaszek** [comment]
@aaloise, I don't have a reproducer. My remark was about a general idea of using any external library that consume `Promise` in its API, so that any client apps cannot do a thing to change a way `Promise` is handled there.

Thanks for raising the issue in Groovy. I hope it's it, but on the other hand I was checking stacktraces for `Promise.future()` and they seemed be the same in cases when code works as expected and when it doesn't. At the very bottom of object creation

**Comment by ptomaszek** [comment]
@aaloise, I don't have a reproducer. My remark was about a general idea of using any external library that consume `Promise` in its API, so that any client app cannot do a thing to change a way `Promise` is handled there.

Thanks for raising the issue in Groovy :+1:

**Comment by ptomaszek** [comment]
I narrowed different Groovy behaviors and I'm sharing it because someone may be faster in finding the exact root cause

1. Groovy 3.0.6, Java 11
1. Calling `Promise.future()` runs through the same stacktrace for valid and invalid results:
   ```
   createPojoMetaMethodSite:296, CachedMethod (org.codehaus.groovy.reflection)
   createCachedMethodSite:162, PojoMetaMethodSite (org.codehaus.groovy.runtime.callsite)
   createPojoMetaMethodSite:151, PojoMetaMethodSite (org.codehaus.groovy.runtime.callsite)
   createPojoCallSite:3525, MetaClassImpl (groovy.lang)
   createPojoSite:123, CallSiteArray (org.codehaus.groovy.runtime.callsite)
   createCallSite:160, CallSiteArray (org.codehaus.groovy.runtime.callsite)
   defaultCall:47, CallSiteArray (org.codehaus.groovy.runtime.callsite)
   call:125, AbstractCallSite (org.codehaus.groovy.runtime.callsite)
   call:130, AbstractCallSite (org.codehaus.groovy.runtime.callsite)
   main:12, MainGroovy (pt)
   ```
1. At some point `MetaClassImpl.getMethodWithCachingInternal()` gets called, exactly here:
   ```
   getMethodWithCachingInternal:1393, MetaClassImpl (groovy.lang)
   createPojoCallSite:3523, MetaClassImpl (groovy.lang)
   ```
1. Debug it and call `((CachedMethod) ((FastArray) e.methods).data[0]).pojoCallSiteConstructor.referent.clazz`
   - when it's `io.vertx.core.Future$future`, you'll end up with **invalid result**
   - when it's `io.vertx.core.Promise$future`, you'll end up with **valid result**
1. Resume the code, you'll eventually go there:
   ```
   createPojoMetaMethodSite:296, CachedMethod (org.codehaus.groovy.reflection)
   createCachedMethodSite:162, PojoMetaMethodSite (org.codehaus.groovy.runtime.callsite)
   ```
   and this gets invoked: `return (CallSite) constr.newInstance(site, metaClass, this, params, constr);`
1. Parameters derived from the non-deterministically created `CachedMethod` are the reason why sometimes there's creation of either `Promise` or `Future`, which further causes why `Promise.future()` returns itself (valid) or a new instance (invalid).

**To sum up** - I believe root cause starts in `MetaClassImpl:1392`: 
`final MetaMethodIndex.Entry e = metaMethodIndex.getMethods(sender, site.getName());`
The value `e.methods` gets randomly created with such or such values. I believe it should be always the same. And because it is not, it randomly determines the final result.

![image](https://user-images.githubusercontent.com/5464274/99259713-ae17fb00-281a-11eb-82b7-2d11d5a37150.png)

**Comment by ptomaszek** [comment]
Groovy `3.0.7` will fix the issue: https://issues.apache.org/jira/browse/GROOVY-9815
Should be available in Maven soon.

Thanks @aaloise :+1: 

![image](https://user-images.githubusercontent.com/5464274/100863368-c469bc00-3494-11eb-9dfc-feea4ff9b8f2.png)

**Comment by aaloise** [comment]
Exactly @ptomaszek! It was really an issue with Groovy. Let's wait for the release of 3.0.7 and then test it. If everything works fine, then I can make a simple PR to upgrade Groovy version in both Vert.x 3.9 and 4.0

**Comment by tsegismont** [comment]
Fixed in #115

**Comment by codeheadcm** [comment]
Hello all! I'm actually running into this issue with ...
 
- Groovy 3.0.8
- Vertx 4.2.1
- JDK 11
- JUnit 5.8.1

 and wanted follow-up on this particular case whether there was a fix or if I'm doing something incorrect...

I'm attaching a Gist containing my test cases and (build.gradle for reference) to display the main issue in my case, which is the inability to use the neat "future().onSuccess { ... }" handling. I've read the bug report threads but I didn't quite find the final outcome of the issues.

https://gist.github.com/codeheadcm/c90dfdc7d2b1bdfc708b7e659c02de95

**Comment by codeheadcm** [comment]
Ah, well, now that I went through it again, I can see that the decision is apparently to wait for Groovy 4.0 which should deliver the proper fix as mentioned in https://github.com/vert-x3/vertx-lang-groovy/issues/117
